src/test/java/org/dom4j/util/FeatureCheckingSAXParserFactory.java

```diff
@@ -0,0 +1,125 @@
+package org.dom4j.util;
+
+import org.aopalliance.intercept.MethodInterceptor;
+import org.springframework.aop.framework.ProxyFactory;
+import org.xml.sax.SAXException;
+import org.xml.sax.SAXNotRecognizedException;
+import org.xml.sax.SAXNotSupportedException;
+
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.parsers.SAXParser;
+import javax.xml.parsers.SAXParserFactory;
+import javax.xml.validation.Schema;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * ATLASSIAN: An implementation of {@link SAXParserFactory} which creates {@link org.xml.sax.XMLReader} objects which fail when
+ * the expected set of parser features are missing.
+ */
+public final class FeatureCheckingSAXParserFactory extends SAXParserFactory {
+    private final SAXParserFactory delegate = new org.apache.xerces.jaxp.SAXParserFactoryImpl();
+
+    @SuppressWarnings("unchecked")
+    private static <T> T createProxy(T delegate, MethodInterceptor advice) {
+        ProxyFactory proxyFactory = new ProxyFactory(delegate);
+        proxyFactory.setProxyTargetClass(true);
+        proxyFactory.addAdvice(advice);
+        return (T) proxyFactory.getProxy();
+    }
+
+    @Override
+    public SAXParser newSAXParser() throws ParserConfigurationException, SAXException {
+        return createProxy(delegate.newSAXParser(), SAXParserAllow());
+    }
+
+    private static MethodInterceptor SAXParserAllow() {
+        return invocation -> {
+            Collection<String> allowedMethods = Arrays.asList("getXMLReader");
+            if (allowedMethods.contains(invocation.getMethod().getName()))
+                return createProxy(invocation.proceed(), XmlReaderAllow());
+            else
+                throw new IllegalStateException("Forbidden method " + invocation.getMethod());
+        };
+    }
+
+    private static final Map<String, Boolean> requiredFeatures = new HashMap<String, Boolean>() {{
```

Comment:
For the general purpose fix, we should split this into REQUIRED_FEATURE_DEFAULTS, and add test-visible methods expectExternalGeneralEntities(), expectExternalParameterEntities(), expectLoadExternalDTD(), and resetExpectedFeatures(), so that we don't need to disable any of the tests, and instead we can just update the test setup for specific tests to enable checking exactly what they need, and nothing that they don't.

Reply:
The tests in the 2.1 stream are pretty different anyway, so if they want to go down this route, it's going to need rework.